src/keyrefresh/participant.rs

```diff
@@ -0,0 +1,1065 @@
+            let auxinfo_outputs = auxinfo::Output::simulate_set(&configs, rng);
+
+            // Make the participants
+            zip(configs, zip(keygen_outputs.clone(), auxinfo_outputs))
+                .map(|(config, (keygen_output, auxinfo_output))| {
+                    let input = Input::new(auxinfo_output, keygen_output)?;
+                    Self::new(sid, config.id(), config.other_ids().to_vec(), input)
+                })
+                .collect::<Result<Vec<_>>>()
+        }
+
+        pub fn initialize_keyrefresh_message(
+            &self,
+            keyrefresh_identifier: Identifier,
+        ) -> Result<Message> {
+            let empty: [u8; 0] = [];
+            Message::new(
+                MessageType::Keyrefresh(KeyrefreshMessageType::Ready),
+                keyrefresh_identifier,
+                self.id(),
+                self.id(),
+                &empty,
+            )
+        }
+    }
+
+    /// Delivers all messages into their respective participant's inboxes.
+    fn deliver_all(
+        messages: &[Message],
+        inboxes: &mut HashMap<ParticipantIdentifier, Vec<Message>>,
+    ) {
+        for message in messages {
+            inboxes
+                .get_mut(&message.to())
+                .unwrap()
+                .push(message.clone());
+        }
+    }
+
+    fn is_keyrefresh_done(quorum: &[KeyrefreshParticipant]) -> bool {
+        for participant in quorum {
+            if *participant.status() != Status::TerminatedSuccessfully {
+                return false;
+            }
+        }
+        true
+    }
+
+    #[allow(clippy::type_complexity)]
+    fn process_messages<R: RngCore + CryptoRng>(
+        quorum: &mut [KeyrefreshParticipant],
+        inboxes: &mut HashMap<ParticipantIdentifier, Vec<Message>>,
+        rng: &mut R,
+    ) -> Option<(usize, ProcessOutcome<Output>)> {
+        // Pick a random participant to process
+        let index = rng.gen_range(0..quorum.len());
+        let participant = quorum.get_mut(index).unwrap();
+
+        let inbox = inboxes.get_mut(&participant.id()).unwrap();
+        if inbox.is_empty() {
+            // No messages to process for this participant, so pick another participant
+            return None;
+        }
+        let message = inbox.remove(rng.gen_range(0..inbox.len()));
+        debug!(
+            "processing participant: {}, with message type: {:?} from {}",
+            &participant.id(),
+            &message.message_type(),
+            &message.from(),
+        );
+        Some((index, participant.process_message(rng, &message).unwrap()))
+    }
+
+    #[cfg_attr(feature = "flame_it", flame)]
+    #[test]
+    fn keyrefresh_always_produces_valid_outputs() -> Result<()> {
+        for size in 2..12 {
+            keyrefresh_produces_valid_outputs(size)?;
+        }
+        Ok(())
+    }
+
+    fn keyrefresh_produces_valid_outputs(quorum_size: usize) -> Result<()> {
```

Comment:
I don't know much about cryptography, so this may already be covered by this test. But should there be a check that the signature generated by the old private key shares equals the signature generated by the new private key shares? Or some general check to ensure the new private key is equivalent to the old private key.

Reply:
Absolutely, this is what this test does. Now also added a double-check of this in `maybe_finish`.